### PR TITLE
Improve airways support

### LINKED ISSUE #925: [REQUEST] Airways are not developped in the A320 FMGC FPLN
**Mod Version**

**Describe the bug**

Entering an airway in the A320 FMGC F-PLN is not functional: only the entry and the endpoint of the airway are displayed in the flightplan

**To Reproduce**

1. GOTO FMGC F-PLN page
2. Select a route using airway from a waypoint
3. Enter VIA / TO
4. GOTO F-PLN Page

**Expected behavior**

When entering an airway, all the waypoints between the entry point and the endpoint of the airway should be added in the flightplan list

**Actual behavior**

Only the entry point and the endpoint of the airway are displayed in the flightplan and the intermediate points do not appear

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by santii90**
Fixes #925

## Summary of Changes
Based on https://github.com/flybywiresim/a32nx/pull/1435 changes I got in contact with the owner (@Lessar277) and started working on implementing the latest approach from WorkingTitle.
- Speed is improved (reduced from 1 minute to around 7-12 seconds for an airway with a lot of waypoints)
- Improved airways page, now it should only show the last waypoint and the airway.
- Fixed issue when trying to add VOR
- Airways now have waypoints that are part of a STAR. **Note: This was only made for arrivals, it may be needed for SID, need testing around this.**

## Screenshots (if necessary)
airways

## References 

Actually I'm looking for good references for this so I can improve the general behavior, please feel free to reach me here on discord with them. 

## Additional context

I did a big implementation from WorkingTitle code and some changes to improve speed. Because of that I expect possible issues. 

I'm working on the implementation for scrolling down on the airways pages.

### Testing it:

- I used navigraph add-on to test it, so maybe test it without it
- Try to do radical changes on the flight plan to be sure that nothing broke. Like: I deleted the last waypoint of the airway and it disappear from the airways page and also from FPLAN, but the others waypoints stayed.
- Try to play around with waypoints that may be part of PROC like STAR or SID, example route that I used for debug: HECA DCT CVO A727 LXR M999 MUVOL DCT LADMA OEJN

Discord username (if different from GitHub): santii#9289

## Know issues
- When adding some airways and a direct at the end for example, you are prompted with the insert page and a lot of DIRECT rows, if you click on insert everything is loaded as expected.
- No more than one page for airways right now. Workaround is select a further waypoint.
- Sometimes after you input the final waypoint for the airway all the info on the row disappear but if you go back, the flight plan was updated and also if you go again to the airways page again ,everything is correctly displayed. 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Benjozork** [comment]
Should #1435 then be closed ?

**Comment by santii90** [comment]
> Should #1435 then be closed ?

Not necessary, we should keep #1435 as an initial implementation that is simpler but brings the functionality.

This PR add a lot more code and may take some time to test it. While #1435 have been already tested.

**Review by Benjozork — CHANGES_REQUESTED** [review]

**Comment by boierito** [comment]
Hi! I want to test this PR, how can i help you with that? It should work with all airways, right? Navigraph support?

File: A32NX/html_ui/Pages/VCockpit/Instruments/A320_Neo/Map/A32NX_MapInstrument.html [hunk 1]
```diff
@@ -23,7 +23,7 @@
 <script type="text/html" import-script="/Pages/VCockpit/Instruments/Shared/FlightElements/WaypointLoader.js"></script>
 <script type="text/html" import-script="/Pages/VCockpit/Instruments/Shared/FlightElements/FlightPlan.js"></script>
 <script type="text/html" import-script="/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js"></script>
-<script type="text/html" import-script="/Pages/VCockpit/Instruments/Shared/FlightElements/Waypoint.js"></script>
+<script type="text/html" import-script="/Pages/VCockpit/Instruments/Airliners/Shared/WT/Waypoint.js"></script>
```
[the remark this answers is not in the document]
Comment (santii90) [reply A32NX/html_ui/Pages/VCockpit/Instruments/A320_Neo/Map/A32NX_MapInstrument.html:26]: Yes I tried to figure out where but I left it to the reviewers, so something like:
/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_Waypoint.js
No clue to be honest, do you guys have any guidelines to follow when adding new scripts?

**Comment by Benjozork** [comment]
> Hi! I want to test this PR, how can i help you with that? It should work with all airways, right? Navigraph support?

Hello, only PRs with the "Ready to Test" tag should be tested. This one is in draft.

**Comment by santii90** [comment]
> Hi! I want to test this PR, how can i help you with that? It should work with all airways, right? Navigraph support?

Yes actually I tested it with navigraph on. One thing that didn't work for me was airways that have the same name but a letter at the end, for example UN304 and UN304G, the last was detected also like UN304G, I need to research more but not sure what behavior is expected in real life in this scenario.

**Comment by boierito** [comment]
> > Hi! I want to test this PR, how can i help you with that? It should work with all airways, right? Navigraph support?
> 
> Yes actually I tested it with navigraph on. One thing that didn't work for me was airways that have the same name but a letter at the end, for example UN304 and UN304G, the last was detected also like UN304, I need to research more but not sure what behavior is expected in real life in this scenario.

For me your PR is not working... I am planning this route: RBO2N RBO UN867 DGO UL176 CNA UT182 KEPER KEPE9W

**Comment by Benjozork** [comment]
@boierito as I told you the PR is in draft. If you want to test it anyway, you'll have to include a detailed explanation of the problem.

**Comment by boierito** [comment]
With last commit it works perfect! with this FPL RBO2N RBO UN867 DGO UL176 CNA UT182 KEPER KEPE9W

**Comment by santii90** [comment]
I'm happy with current state. There are more things that I want to implement, like the INSERT yellow menu before actually adding every waypoint of the airway. But I prefer to actually do that on another PR.
For things like adding only the airways names and the MCDU adding the intersection waypoint between both, it's possible because the data is around, but it's a long implementation.

**Comment by danbw9** [comment]
I have been testing this a little bit and this is a great added feature with some minor limitations. some airways and waypoints it will not recognise but pretty much 90% of the time it will work correctly. 

I understand the 'ready to test' label has not been added, but was asked to check this out. Will write a full report when its fully complete.

Thanks Santi for your work. saves people a ton of time!

**Comment by santii90** [comment]
@danbw9 please feel free to comment with the waypoints/airways that are not working for you (if you have the whole route is better) so I can take a look. Also knowing if you had Navigraph plugin is interesting too. Thanks for you time!

**Comment by danbw9** [comment]
@santii90 This route from EGGW to EKCH

MATCH Q295 SOMVA DCT MAVAS DCT DHE DCT DOSUR P729 TUDLO

If I recall correctly, P729 to TUDLO could not input...I am using latest Navigraph beta though.

Check it out and let me know!

**Comment by Carter633** [comment]
Name: British Avgeek YT -- Carter633 #1436
Date of testing: 24/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1460 
Object of test: Check added airway function, loading speed and ability to input different flight plans with airways in MCDU

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
*This is Test 1 of a few I will run on different routes on VATSIM to see how it goes

1. Spawned Cold and Dark at Heathrow for a full flight to Munich 
2. Used SIMBRIEF to create flight plan
3. Inputting all flight plan info into MCDU using Airways page
4. WOW! It makes life a lot quicker on the sim!
5. Worked with SID and STAR for me out of Heathrow and into Munich

Flight notes
For reference: Route used is as follows ---
DET2F DET L6 DVR UL9 KONAN UL607 MATUG DCT MOVUM T109 HAREM T104 ANORA ANOR3A

Startup: Everything is normal

Taxi/Takeoff : Had a couple of very quick Direct To from ATC - didn't create any issues at all with this. SID fed into the L6 airway - no issues at all during input!

Cruise: Everything is normal including a few more "DCT to" from ATC

Descent/Landing : STAR was accurate and flowed in from the T104 Airway perfectly

Shutdown: N/A

Issues (If Any): None found in Test 1
Overall Rating: Good
Conclusions of the testing: Really impressed with this. Used Simbrief and as mentioned above, had a STAR that fed in from the T104 airway and that worked absolutely fine. Huge improvement - before this, I was planning in the World Map (sim main menu!) as it was a frustration to do it in the MCDU on stand but now this is a win! Thanks for the great work.

Will run a few more tests tomorrow on some more full flights to see if any other issues crop up.

File: A32NX/html_ui/Pages/VCockpit/Instruments/A320_Neo/Map/A32NX_MapInstrument.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (lousybyte) [reply A32NX/html_ui/Pages/VCockpit/Instruments/A320_Neo/Map/A32NX_MapInstrument.html]: I think the convention so far is to prefix the modified file names with A32NX_ and adjust those in every HTML files that include them. 
Check [https://github.com/flybywiresim/a32nx/commit/848de876f334dcc32ec79307b24876af42330e6c#diff-1a4f0f0dbbd73af4b8e2bea0c691cfc053b7d51808c4f3017d31003f39f8db35](https://github.com/flybywiresim/a32nx/commit/848de876f334dcc32ec79307b24876af42330e6c#diff-1a4f0f0dbbd73af4b8e2bea0c691cfc053b7d51808c4f3017d31003f39f8db35)
Can argue as mentioned above they could be just be put in a A32NX folder while keeping the original names.

It's also preferably to keep the file include commit on its own, so any changes to them are easily visible in separate commits.
This is only valid if there are actually any changes to those new files, you don't need to include them if they are the default ones.

**Comment by danbw9** [comment]
Name: danbw9#7028
Date of testing: 24/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1460
Object of test: Check the enhancement of the airways functionality through MCDU. 

Testing technique: Evaluate new features compared to previous behaviour.
Description of how you tested the Pull Request: Complete normal flight as normal.
Steps you've made to test the Pull Request: 

C&D for flight from EGLL to EKCH . Did normal setup for normal flight. In the route input, double checked all the airways and intersections are correctly put in and then checking all the waypoints are correct and look for any lost waypoints as was happening before. Checked in CRZ and check via the PLAN page on ND and via F PLAN page via MCDU. 

Flight notes. I tested this PR a few days ago and had an issue with a specific route. I have since flown that same route and that has been resolved. Lower airways were causing some visual issues, but functionality wise all okay.

Startup: Everything is normal. Checked through route in the pre flight and the airways I entered. Cross checked with OFP

Taxi/Takeoff : N/A
Cruise: Everything is normal. Tried to enter a new airway in flight and it was not temperamental with it

Descent/Landing : P729 to TUDLO all good.

Shutdown: N/A

Issues (If Any): N/A
Overall Rating: Good
Conclusions of the testing: This is a game changer if you input your route manually. Saves a lot of time. Good job Santi.

**Comment by Carter633** [comment]
Name: British Avgeek YT -- Carter633 #1436
Date of testing: 25/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1460
Object of test: Check added airway function, loading speed and ability to input different flight plans with airways in MCDU

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
*This is Test 2 of a few I will run on different routes on VATSIM to see how it goes

1. Spawned Cold and Dark at London Gatwick EGKK for a full flight to Skiathos Island, Greece LGSK (VORDME only approach)
2. Used SIMBRIEF to create flight plan
3. Inputting all flight plan info into MCDU using Airways page mixed with a few DCT to waypoints within the plan
4. Checked function surrounding SIDs and STARs though LGSK had a direct to the DME VOR approach

Flight notes
For reference: Route used is as follows ---
BOGN1M BOGNA L612 XAMAB/N0455F350 UL612 MADIV/N0444F370 UL612 BIBAN DCT PAPIZ DCT PINDO UL607 PIKOS UM600 KOGIS/N0349F130 W58 SKP DCT

Startup: Everything is normal - input was a breeze! Really good, loving it - second test and I'm still so so pleased I can do this now!

Taxi/Takeoff : Everything behaved as normal. This test I was able to just fly the planned SID into L612

Cruise: Everything is normal

Descent/Landing :  Airway dropped straight into a DCT to the airport via the VOR DME approach and this did not cause any issues

Shutdown: N/A

Issues (If Any): None found in Test 2
Overall Rating: Good
Conclusions of the testing: Used Simbrief and as mentioned above, Contrast to the first test which used a STAR and ILS approach into Munich, I flew this using a VOR DME into a small greek island with the approach dropping straight out from the airway onto a VOR DME. Same positive comments as test 1 - Huge improvement and is a huge game changer for the A32nx in the sim!

**Review by Benjozork — APPROVED** [review]

**Pull request merged**